### PR TITLE
[WIP] Added crossdock integration tests

### LINKED ISSUE #85: Create crossdock integration tests
To address issues like https://github.com/jaegertracing/jaeger-client-csharp/issues/80#issuecomment-394416348

### PR DESCRIPTION
**Pull request by Falco20019**
## Which problem is this PR solving?
- Fixes #85 

## Short description of the changes
- Added crossdock support to jaeger-client-csharp
- Testing against csharp, go, java and python

### CODE REVIEW

**Comment by Falco20019** [comment]
@yurishkuro I just added full crossdock support (client + server). I just have to find out how to generate the xdock-csharp and also how to publish it automatically through appveyor. Can you please have a look at the docker-compose file if it looks correct to you?

**Comment by yurishkuro** [comment]
docker-compose looks fine.  The only think I would suggest considering is not importing `jaeger-docker-compose.yml` and instead using jaeger-all-in-one container, because the former requires Cassandra, while the latter is just one extra container, much more lightweight and more reliable in testing (Cassandra takes a while to start). Functionally, it doesn't matter much, specifically for the client testing. I would've liked to have all clients' crossdock tests to switch to all-in-one, but so far there's no precedent.

**Comment by Falco20019** [comment]
Thanks for the note, I will try to adjust it and add the necessary environment variables to the docker compose file so that all xdock endpoints are able to use the all-in-one package. As far as I have seen, at least go and java kept it configurable.

**Comment by Falco20019** [comment]
@yurishkuro I adjusted the files to use `jaegertracing/all-in-one`.

I was able to get everyone talking to the service using links. `xdock-java` seems to have a bug and uses default UdpSender in some places, so setting the environment `AGENT_HOST_PORT=jaeger:5775` is not working all the way through...

Only the `test_driver` is not coming up and is reporting that the `jaeger-query` port `16687` (`QueryHTTPAdmin`) is not reachable. I tried forwarding it to `localhost` but can't call it manually either. Is this port somehow blocked/modified in `all-in-one` so that I need to set something?

Attached the logs for `test_driver` and `jaeger`:
```
C:\Users\xxx\Documents\GitHub\jaeger-client-csharp\crossdock>docker-compose logs test_driver
Attaching to crossdock_test_driver_1
test_driver_1  | 2019-07-16T14:36:53.466Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:36:54.472Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:36:55.482Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:36:56.484Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:36:57.488Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:36:58.489Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:36:59.491Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:37:00.492Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:37:01.494Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:37:02.496Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}
test_driver_1  | 2019-07-16T14:37:03.497Z       INFO    crossdock/main.go:116   Health check failed     {"service": "jaeger-query", "error": "Get http://jaeger-query:16687: dial tcp 172.23.0.2:16687: connect: connection refused"}

C:\Users\xxx\Documents\GitHub\jaeger-client-csharp\crossdock>docker-compose logs jaeger
Attaching to crossdock_jaeger_1
jaeger_1       | 2019/07/16 14:36:49 maxprocs: Leaving GOMAXPROCS=2: CPU quota undefined
jaeger_1       | {"level":"info","ts":1563287809.6383483,"caller":"flags/service.go:115","msg":"Mounting metrics handler on admin server","route":"/metrics"}
jaeger_1       | {"level":"info","ts":1563287809.6391206,"caller":"flags/admin.go:108","msg":"Mounting health check on admin server","route":"/"}
jaeger_1       | {"level":"info","ts":1563287809.639397,"caller":"flags/admin.go:114","msg":"Starting admin HTTP server","http-port":14269}
jaeger_1       | {"level":"info","ts":1563287809.63953,"caller":"flags/admin.go:100","msg":"Admin server started","http-port":14269,"health-status":"unavailable"}
jaeger_1       | {"level":"info","ts":1563287809.7325983,"caller":"memory/factory.go:55","msg":"Memory storage initialized","configuration":{"MaxTraces":0}}
jaeger_1       | {"level":"info","ts":1563287809.750669,"caller":"grpc/builder.go:75","msg":"Agent requested insecure grpc connection to collector(s)"}
jaeger_1       | {"level":"info","ts":1563287809.7507243,"caller":"grpc/clientconn.go:242","msg":"parsed scheme: \"\"","system":"grpc","grpc_log":true}
jaeger_1       | {"level":"info","ts":1563287809.7507558,"caller":"grpc/clientconn.go:248","msg":"scheme \"\" not registered, fallback to default scheme","system":"grpc","grpc_log":true}
jaeger_1       | {"level":"info","ts":1563287809.7507794,"caller":"grpc/resolver_conn_wrapper.go:113","msg":"ccResolverWrapper: sending update to cc: {[{127.0.0.1:14250 0  <nil>}] }","system":"grpc","grpc_log":true}
jaeger_1       | {"level":"info","ts":1563287809.7510219,"caller":"base/balancer.go:76","msg":"base.baseBalancer: got new resolver state: {[{127.0.0.1:14250 0  <nil>}] }","system":"grpc","grpc_log":true}
jaeger_1       | {"level":"info","ts":1563287809.751093,"caller":"base/balancer.go:130","msg":"base.baseBalancer: handle SubConn state change: 0xc0004760a0, CONNECTING","system":"grpc","grpc_log":true}
jaeger_1       | {"level":"info","ts":1563287809.7516277,"caller":"all-in-one/main.go:198","msg":"Starting agent"}
jaeger_1       | {"level":"info","ts":1563287809.7518241,"caller":"grpc/clientconn.go:1139","msg":"grpc: addrConn.createTransport failed to connect to {127.0.0.1:14250 0  <nil>}. Err :connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:14250: connect: connection refused\". Reconnecting...","system":"grpc","grpc_log":true}
jaeger_1       | {"level":"info","ts":1563287809.7518706,"caller":"base/balancer.go:130","msg":"base.baseBalancer: handle SubConn state change: 0xc0004760a0, TRANSIENT_FAILURE","system":"grpc","grpc_log":true}
jaeger_1       | {"level":"info","ts":1563287809.7522886,"caller":"app/agent.go:68","msg":"Starting jaeger-agent HTTP server","http-port":5778}
jaeger_1       | {"level":"info","ts":1563287809.7579002,"caller":"all-in-one/main.go:241","msg":"Starting jaeger-collector TChannel server","port":14267}
jaeger_1       | {"level":"info","ts":1563287809.7579823,"caller":"grpcserver/grpc_server.go:64","msg":"Starting jaeger-collector gRPC server","grpc-port":"14250"}
jaeger_1       | {"level":"info","ts":1563287809.75808,"caller":"all-in-one/main.go:259","msg":"Starting jaeger-collector HTTP server","http-port":14268}
jaeger_1       | {"level":"info","ts":1563287809.783146,"caller":"querysvc/query_service.go:130","msg":"Archive storage not created","reason":"archive storage not supported"}
jaeger_1       | {"level":"info","ts":1563287809.7831776,"caller":"all-in-one/main.go:341","msg":"Archive storage not initialized"}
jaeger_1       | {"level":"info","ts":1563287809.7841458,"caller":"healthcheck/handler.go:129","msg":"Health Check state change","status":"ready"}
jaeger_1       | {"level":"info","ts":1563287809.784173,"caller":"app/server.go:134","msg":"Starting CMUX server","port":16686}
jaeger_1       | {"level":"info","ts":1563287809.7841902,"caller":"app/server.go:111","msg":"Starting HTTP server","port":16686}
jaeger_1       | {"level":"info","ts":1563287809.7842042,"caller":"app/server.go:124","msg":"Starting GRPC server","port":16686}
jaeger_1       | {"level":"info","ts":1563287809.8135185,"caller":"all-in-one/main.go:297","msg":"Listening for Zipkin HTTP traffic","zipkin.http-port":9411}
jaeger_1       | {"level":"debug","ts":1563288029.7370896,"caller":"processors/thrift_processor.go:115","msg":"Span(s) received by the agent","bytes-received":644}
jaeger_1       | {"level":"debug","ts":1563288029.7478309,"caller":"app/span_processor.go:117","msg":"Span written to the storage by the collector","trace-id":"14761844f41a67a4","span-id":"14761844f41a67a4"}
jaeger_1       | {"level":"debug","ts":1563288029.7478356,"caller":"app/span_processor.go:117","msg":"Span written to the storage by the collector","trace-id":"7325878b82f56001","span-id":"7325878b82f56001"}
```

**Comment by yurishkuro** [comment]
You can remove Java for now if it's not configurable (although it's surprising).

We may need to extend test_driver to support extra configuration. But could you try naming all-in-one container as `jaeger-agent`?

**Comment by Falco20019** [comment]
With links (and aliasing), Java works. But not just with environment variable as it should. I will create a PR in Java for that once this is working. Only problem still around is that test_driver wants to connect to jaeger-query:16687 which is strangely not available (See logs). The logs of Jaeger also don’t show that there should be something on this port. I found in the code for query that this is the AdminHTTP, but don’t know the code base of query and all-in-one to know why it’s not there. The port mapping in the compose file is there.

**Comment by yurishkuro** [comment]
> Only problem remaining is that test_driver wants to connect to jaeger-query:16687 which is strangely not available

that's understandable, I think the driver hits the health endpoints on all services to ensure they are ready before running the tests, but with all-in-one there  is only one health endpoint. We can probably re-map the 3 admin ports to a single port inside all-in-one to make driver happy.

File: crossdock/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            - "8080-8082"
+
+    test_driver:
+        image: jaegertracing/test-driver
+        depends_on:
+            - jaeger
+        links:
+            - "jaeger:jaeger-agent"
+            - "jaeger:jaeger-collector"
+            - "jaeger:jaeger-query"
```
Comment (yurishkuro) [inline crossdock/docker-compose.yml]: nice!

**Comment by Falco20019** [comment]
@yurishkuro I checked the code that `test_driver` is running and found that the endpoint for the query health check is configurable through `JAEGER_QUERY_HC_HOST_PORT`. So I set it to `jaeger-query:14269` and it's running like a charme now :) So it's basically checking the only existing health endpoint twice now and making that configurable would be the nicer option, but at least we could use this as a working solution in our build chain right now.

I will try to get it into appveyor and let @saschaholesch review everything if it sounds good from your side.

**Comment by Falco20019** [comment]
Sadly it's not easily solvable with Appveyor using the Windows Environment (what we use now). The crossdock containers are all Linux containers and therefore we can not run them on our environment without nested virtualization (that we don't have).

I will look into switching our environment to Ubuntu and might create a new PR afterwards. The crossdock code itself should work just fine though, so I'm keeping the branch around.

**Pull request closed without merging**